moonraker/components/wled.py

```diff
@@ -0,0 +1,502 @@
+        self.send_full_chain_data = True
+        self.onoff = OnOff.on
+        self.preset = self.initial_preset
+        if self.initial_preset >= 0:
+            self._update_color_data(self.initial_red,
+                                    self.initial_green,
+                                    self.initial_blue,
+                                    self.initial_white)
+            await self.wled_on(self.initial_preset)
+        else:
+            await self.set_wled(self.initial_red,
+                                self.initial_green,
+                                self.initial_blue,
+                                self.initial_white)
+
+    def _update_color_data(self, red, green, blue, white, index=None):
+        red = int(red * 255. + .5)
+        blue = int(blue * 255. + .5)
+        green = int(green * 255. + .5)
+        white = int(white * 255. + .5)
+        if self.color_order is ColorOrder.RGB:
+            led_data = [red, green, blue]
+        else:
+            led_data = [red, green, blue, white]
+
+        if index is None:
+            self._chain_data[:] = led_data * self.chain_count
+        else:
+            elem_size = len(led_data)
+            self._chain_data[(index-1)*elem_size:index*elem_size] = led_data
+
+    async def _send_wled_command(self,
+                                 state: Dict[str, Any]):
+        async with self.request_mutex:
+            try:
+                if not hasattr(self, "wled_info"):
+                    state['v'] = True
+
+                logging.debug(f"WLED: url:{self.url} json:{state}")
+
+                headers = {"Content-Type": "application/json"}
+                request = HTTPRequest(url=self.url,
+                                      method="POST",
+                                      headers=headers,
+                                      body=json.dumps(state),
+                                      connect_timeout=self.timeout,
+                                      request_timeout=self.timeout)
+                response = await self.client.fetch(request)
+
+                logging.debug(
+                    f"WLED: url:{self.url} status:{response.code} "
+                    f"response:{response.body}")
+
+                # Generally ignore response unless fetching information
+                if not hasattr(self, "wled_info"):
+                    data = json_decode(response.body)
+
+                    logging.debug(f"WLED: url:{self.url} data:{data}")
+
+                    self.wled_info = data['info']
+                    self.wled_effects = [e.lower() for e in data['effects']]
+                    self.wled_palettes = [p.lower() for p in data['palettes']]
+
+                self.error_state = None
+            except Exception as e:
+                msg = f"WLED: Error {e}"
+                self.error_state = msg
+                logging.exception(msg)
+                raise self.server.error(msg)
+
+    async def wled_on(self, preset):
```

Comment:
The argument and return value needs an annotation.

Reply:
Believed fixed